### PR TITLE
Data Protection Working Group

### PR DESCRIPTION
**Pull request by xing-yang**
This PR proposes to form a Data Protection Working Group.  This is a cross SIG collaboration between SIG-Apps and SIG-Storage.

Discussed at SIG-Apps and SIG-Storage meetings and proposed in the mailing list:
https://groups.google.com/forum/#!topic/kubernetes-dev/0ZWLfML3uS0

committee/steering
Approvers: @saad-ali @prydonius

### CODE REVIEW

**Comment by nikhita** [comment]
/committee steering

cc @kubernetes/steering-committee

File: wg-data-protection/charter.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+this requires.
+
+The Data Protection Working Group is organized with the goal of providing
+a cross SIG forum to discuss how to support data protection in Kubernetes,
+identify missing functionality, and work together to design features that
+are needed to achieve the goal.
+
+### In scope
+
+* High level discussions on what it means to support data protection in Kubernetes at different level and how to do it.
```
Comment (prydonius) [inline wg-data-protection/charter.md]:
```suggestion
* High level discussions on what it means to support data protection in Kubernetes at different levels and how to do it.
```

File: wg-data-protection/charter.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+* Design discussions on specific topics related to data protection and disaster recovery support.
+* Document results of discussions and investigations in a linkable medium.
+
+Potential design topics include, but are not limited to the following:
+* Extract snapshot data without cloning
+* Volume backups
+* Data populator
+* Retrieve diffs between two snapshots (block and file level)
+* Consistency volume groups (group snapshot)
```
Comment (kmova) [inline wg-data-protection/charter.md]: Does the scope for group snapshots include - volumes potentially across different pods and namespaces?

Comment (yuxiangqian) [reply wg-data-protection/charter.md]: @kmova those are the details the WG is looking to discuss and define.

**Comment by yuxiangqian** [comment]
cc @jingxu97

**Comment by yuxiangqian** [comment]
cc @alarge

**Comment by xing-yang** [comment]
cc @liyinan926

File: wg-data-protection/charter.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+Data protection in Kubernetes context typically involves backup and recovery
+of two types of entities:
+* Kubernetes object resources
+* Persistent volume data
+We consider it a complicated and layered problem, including backup and recovery
+at persistent volume level, application level, and cluster level. Part of the
+working group’s charter is to define what Kubernetes native constructs are
+required to achieve these goals.
+
+Data Protection in Kubernetes includes backup and restore of applications
```
Comment (jingxu97) [inline wg-data-protection/charter.md]: Above it mention three levels. Here it only talks about application restore. Does it need to be mentioned separately?

File: wg-data-protection/charter.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+are needed to achieve the goal.
+
+### In scope
+
+* High level discussions on what it means to support data protection in Kubernetes at different level and how to do it.
+* Design discussions on specific topics related to data protection and disaster recovery support.
+* Document results of discussions and investigations in a linkable medium.
+
+Potential design topics include, but are not limited to the following:
+* Extract snapshot data without cloning
```
Comment (jingxu97) [inline wg-data-protection/charter.md]: what does this mean?

File: wg-data-protection/charter.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+* Volume backups
+* Data populator
+* Retrieve diffs between two snapshots (block and file level)
```
Comment (jingxu97) [inline wg-data-protection/charter.md]: it seems a very low level functionality. Should it be part of kubernetes function?

File: wg-data-protection/charter.md (hunk 3 above)
Comment (xing-yang) [reply wg-data-protection/charter.md]: Hmm, it looks like I forgot to delete this line.  Good catch.

File: wg-data-protection/charter.md (hunk 4 above)
Comment (xing-yang) [reply wg-data-protection/charter.md]: It means the ability to read data from a snapshot without creating a new volume.

File: wg-data-protection/charter.md (hunk 5 above)
Comment (xing-yang) [reply wg-data-protection/charter.md]: Without an kubernetes API to expose this, there is no easy way for backup vendor to retrieve this.  Lots of backup vendors requested this feature.  We should think about how to do it.

File: wg-data-protection/charter.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+
+## Scope
+
+Data protection in Kubernetes context typically involves backup and recovery
+of two types of entities:
+* Kubernetes object resources
+* Persistent volume data
+We consider it a complicated and layered problem, including backup and recovery
+at persistent volume level, application level, and cluster level. Part of the
```
Comment (kmova) [inline wg-data-protection/charter.md]: what does `application` mean? Is it a kubernetes object like - a StatefulSet, Deployment, or does it include a collection of objects (deployments, statefulsets and daemonsets) that together form an application? 

Probably, `complicated and layered problem` means just that - it can be volume, application or cluster and also combination of multiple applications.

Comment (prydonius) [reply wg-data-protection/charter.md]: I believe this is related to https://github.com/kubernetes/enhancements/pull/1051, so a collection of objects that form an application.

Comment (yuxiangqian) [reply wg-data-protection/charter.md]: @kmova that's a really interesting point.
as @prydonius commented, while there is an ongoing KEP from @liyinan926 focusing on statefulset. there also have been very initial discussions around how to backup/restore a set of objects which form an application consistently (i.e., a wordpress application).
Again, those are exactly the topics the proposed WG try to collect thoughts, organize discussions, and come up with a definition.

Comment (liyinan926) [reply wg-data-protection/charter.md]: Just to clarify, the KEP https://github.com/kubernetes/enhancements/pull/1051 focuses on the APIs for defining and performing data protection operations on stateful applications, which consist of a collection of related Kubernetes resources (including PVCs) that make up the applications.

File: wg-data-protection/charter.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+* Design discussions on specific topics related to data protection and disaster recovery support.
+* Document results of discussions and investigations in a linkable medium.
+
+Potential design topics include, but are not limited to the following:
+* Extract snapshot data without cloning
+* Volume backups
+* Data populator
+* Retrieve diffs between two snapshots (block and file level)
+* Consistency volume groups (group snapshot)
+* Application snapshot
```
Comment (liyinan926) [inline wg-data-protection/charter.md]: Suggested changing to `Application snapshot, backup, and recovery`

File: wg-data-protection/charter.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+* Data protection policy
+* Data protection workflows
+
+### Out of Scope
+
+* Design discussions not related to data protection is out of scope.
```
Comment (spiffxp) [inline wg-data-protection/charter.md]: Can you be more specific? 

For example, what sorts of design discussions would belong better specifically within SIG Apps or SIG Storage independent of this WG?

File: wg-data-protection/charter.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+* This is a working group so it does not own code. Design discussions for
+  a specific feature can happen in the working group but code implementation
+  will be owned by SIG-Apps or SIG-Storage.
+
+
+## Stakeholders
+
+Stakeholders for this working group include members in the following SIGs:
+* SIG APIs
```
Comment (spiffxp) [inline wg-data-protection/charter.md]: I think you mean SIG Apps?

File: wg-data-protection/charter.md (hunk 7 above)
Comment (xing-yang) [reply wg-data-protection/charter.md]: Sure.

File: wg-data-protection/charter.md (hunk 9 above)
Comment (xing-yang) [reply wg-data-protection/charter.md]: You are right!

File: wg-data-protection/charter.md (hunk 8 above)
Comment (xing-yang) [reply wg-data-protection/charter.md]: For example, how to enhancement volume placement and how to report volume health belong to SIG Storage and would not be a focus area of this WG.  Workload API designs for StatefulSet and Deployment belong to SIG Apps, however, this WG would be interested in figuring out how to backup and recover a StatefulSet or Deployment.

Comment (spiffxp) [reply wg-data-protection/charter.md]: thanks

**Comment by spiffxp** [comment]
LGTM from steering member, need three more

File: wg-data-protection/charter.md (hunk 1 above)
Comment (xing-yang) [reply wg-data-protection/charter.md]: Done

File: sigs.yaml [hunk 10]
```diff
@@ -2306,6 +2306,39 @@ workinggroups:
...
+      name: Xiangqian Yu
+      company: Google
+  meetings:
+  - description: Regular WG Meeting
+    day: Monday
+    time: "9:00"
+    tz: PT (Pacific Time)
+    frequency: bi-weekly
+    url: https://docs.google.com/document/d/15tLCV3csvjHbKb16DVk-mfUmFry_Rlwo-2uG6KNGsfw/edit
+    archive_url: http://bit.ly/wg-data-protection-notes
```
Comment (dims) [inline sigs.yaml]: 404, probably have to fix this when we have a real url i guess

File: sigs.yaml [hunk 11]
```diff
@@ -2306,6 +2306,39 @@ workinggroups:
...
+    recordings_url: http://bit.ly/data-protection-playlist
```
Comment (dims) [inline sigs.yaml]: same as above

File: sigs.yaml [hunk 12]
```diff
@@ -2306,6 +2306,39 @@ workinggroups:
...
+  contact:
+    slack: wg-data-protection
+    mailing_list: https://groups.google.com/forum/#!forum/kubernetes-data-protection
```
Comment (dims) [inline sigs.yaml]: You can probably create the google group yourself, just make sure all the leads are admins.

File: wg-data-protection/charter.md [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+* Data populator
+* Retrieve diffs between two snapshots (block and file level)
+* Consistency volume groups (group snapshot)
+* Application snapshot, backup, and recovery
+* Data protection policy
+* Data protection workflows
+
+### Out of Scope
+
+* Design discussions not related to data protection is out of scope. For example,
```
Comment (dims) [inline wg-data-protection/charter.md]: thanks for adding good notes here for out of scope :)

File: wg-data-protection/charter.md [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+  how to enhance volume placement and how to report volume health belong to SIG
+  Storage and would not be a focus area of this WG. Workload API designs for
+  StatefulSet and Deployment belong to SIG Apps, however, this WG would be interested
+  in figuring out how to backup and recover a StatefulSet or Deployment.
+* This is a working group so it does not own code. Design discussions for
+  a specific feature can happen in the working group but code implementation
```
Comment (dims) [inline wg-data-protection/charter.md]: want to tweak a bit to say "code implementation / KEPs"?

File: wg-data-protection/charter.md [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+  will be owned by SIG-Apps or SIG-Storage.
+
+
+## Stakeholders
+
+Stakeholders for this working group include members in the following SIGs:
+* SIG Apps
+* SIG Storage
+
+Stakeholders also include backup vendors who want to provide data protection
```
Comment (dims) [inline wg-data-protection/charter.md]: will sig-auth be needed? (is there some security aspect? encrypted volumes maybe?)

File: wg-data-protection/charter.md (hunk 14 above)
Comment (xing-yang) [reply wg-data-protection/charter.md]: We can still review KEPs in the WG as we would be interested in the design, but the approval of the KEPs belong to the SIGs?

File: wg-data-protection/charter.md (hunk 15 above)
Comment (xing-yang) [reply wg-data-protection/charter.md]: Sure.  Will add sig-auth.

File: sigs.yaml (hunk 12 above)
Comment (xing-yang) [reply sigs.yaml]: Sure

File: sigs.yaml (hunk 10 above)
Comment (xing-yang) [reply sigs.yaml]: I'm not sure what is the process of creating this url.  Will this be provided after the WG is approved?   Sure we can fix it after the url becomes available.

File: wg-data-protection/charter.md (hunk 14 above)
Comment (dims) [reply wg-data-protection/charter.md]: yep

File: sigs.yaml (hunk 12 above)
Comment (xing-yang) [reply sigs.yaml]: WIP

File: wg-data-protection/charter.md (hunk 14 above)
Comment (xing-yang) [reply wg-data-protection/charter.md]: Fixed.

File: wg-data-protection/charter.md (hunk 15 above)
Comment (xing-yang) [reply wg-data-protection/charter.md]: Added sig-auth.

**Comment by dims** [comment]
LGTM from me (any nits from me are not blocking)

File: wg-data-protection/charter.md [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+* SIG Storage
+* SIG Auth (consult SIG Auth from security aspect)
+
+Stakeholders also include backup vendors who want to provide data protection
+support in Kubernetes and end users who want to use data protection applications.
+
+
+## Disband criteria
+
+It is our intent to disband once data protection support is mature in
```
Comment (nrb) [inline wg-data-protection/charter.md]: From the wg-governance document:

>How does the group know when the problem solving process is completed, and it is time for the Working Group to dissolve?

I think I would ask here, what does "mature" mean in this context?

Comment (xing-yang) [reply wg-data-protection/charter.md]: I think "mature" means people in the WG believe there is no more missing functionality for Data Protection in Kubernetes and the Data Protection workflow is well defined and well supported.

File: sigs.yaml (hunk 12 above)
Comment (xing-yang) [reply sigs.yaml]: Fixed.

File: sigs.yaml [hunk 17]
```diff
@@ -2306,6 +2306,39 @@ workinggroups:
...
+  mission_statement: >
+    A Working Group dedicated to promoting data protection support in Kubernetes,
+    identifying missing functionality and working together to design features to enable
+    data protection support. Involves collaboration with multiple SIGs such as Apps
+    and Storage.
+
+  charter_link: charter.md
+  stakeholder_sigs:
+  - Apps
+  - Storage
```
Comment (nikhita) [inline sigs.yaml]: I think Auth should be added here as well.

Comment (xing-yang) [reply sigs.yaml]: Will do.  Thanks!

File: wg-data-protection/charter.md (hunk 16 above)
Comment (derekwaynecarr) [reply wg-data-protection/charter.md]: It would be good if we could crisp up the statement.  It seems like the WG wants to establish a cross-SIG roadmap for a set of enhancements required to meet the identified requirements.  Is it possible to say that the WG will define that document, and will disband once the roadmap is agreed upon with each of the SIG stakeholders?  At that point, its a matter of execution against the roadmap which should happen with the code owning bodies (i.e. SIGS)?  I am trying to think about how to use a crisper definition of mature oriented around a work product that is not code.

**Comment by enj** [comment]
How exactly does this relate to Auth?  I see that `Data protection policy` is listed, which I take to mean "backups must be encrypted" and "secrets must be encrypted at rest and in transit."  The word "policy" is incredibly broad and vague when applied to Kube, so I would like to be a bit more explicit on what is meant here.  For example, "secrets must be encrypted at rest and in transit" is generally owned by Auth and API machinery.

**Comment by xing-yang** [comment]
> How exactly does this relate to Auth? I see that Data protection policy is listed, which I take to mean "backups must be encrypted" and "secrets must be encrypted at rest and in transit." The word "policy" is incredibly broad and vague when applied to Kube, so I would like to be a bit more explicit on what is meant here. For example, "secrets must be encrypted at rest and in transit" is generally owned by Auth and API machinery.

I added SIG-Auth as a stakeholder SIG after getting a review comment that there are security aspects that this WG needs to consider. I think every SIG and WG in Kubernetes may need to consult SIG-Auth on security related issues.  I’m actually not sure whether SIG-Auth should be listed explicitly in the charter for this WG.  I’m open to suggestions from SIG-Auth chairs.

Regarding Data Protection Policy, it usually means we can set up a schedule to do periodic backups, set a backup retention policy to automatically clean up old backups, set a topology to specify a backup location, etc.  It can possibly include policies such as “backups must be encrypted” and “secrets must be encrypted at rest and in transit”.  As a WG, we’ll discuss more on what should be covered there and how to achieve that.

**Comment by xing-yang** [comment]
> It would be good if we could crisp up the statement. It seems like the WG wants to establish a cross-SIG roadmap for a set of enhancements required to meet the identified requirements. Is it possible to say that the WG will define that document, and will disband once the roadmap is agreed upon with each of the SIG stakeholders? At that point, its a matter of execution against the roadmap which should happen with the code owning bodies (i.e. SIGS)? I am trying to think about how to use a crisper definition of mature oriented around a work product that is not code.

I don’t think a WG can establish a cross-SIG roadmap as it does not own code?  I also don’t know if a SIG can set a roadmap.  In SIG-Storage, we track work items for each Kubernetes release.   We prioritize work times in each release.  If an item is not completed as planned for a given release, it will be moved to the next release.  I have not seen a one year or multi-year roadmap there.

Let me try to add some clarifications for disband criteria.

**Comment by xing-yang** [comment]
@enj and @derekwaynecarr, addressed your comments.  PTAL.  Thanks.

**Comment by enj** [comment]
I am okay with SIG Auth not being explicitly listed as a stakeholder here (though I would like opinions from the other tech leads).  We try to avoid being SIG "Security" since everything has some security aspect.

**Comment by xing-yang** [comment]
> I am okay with SIG Auth not being explicitly listed as a stakeholder here (though I would like opinions from the other tech leads). We try to avoid being SIG "Security" since everything has some security aspect.

Thanks @enj. Either way I'm open to suggestions from SIG Auth leads.

**Comment by xing-yang** [comment]
Talked to @deads2k and he thinks that SIG-Auth does not need to be an endorsing SIG, but the WG will need to engage to some degree with SIG-Auth.

So I removed SIG Auth as an endorsing SIG and stated that we will consult SIG Auth from security aspects.

File: wg-data-protection/charter.md [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+the general conventions described in the [Kubernetes Charter README] and
+the Roles and Organization Management outlined in [sig-governance], where
+applicable to a Working Group.
+
+
+## Scope
+
+Data protection in Kubernetes context typically involves backup and recovery
+of two types of entities:
+* Kubernetes API object resources
```
Comment (mikedanese) [inline wg-data-protection/charter.md]: The term "data protection" is typically associated with information security and privacy. It sounds like the goals of this working group are not related. Is there a better term we could use to summarize the work here?

Comment (xing-yang) [reply wg-data-protection/charter.md]: "Data protection" is typically the term used by backup vendors.  I'll see if I can add some clarification here.

Comment (mikedanese) [reply wg-data-protection/charter.md]: It definitely is a matter of perspective. I immediately thought of GDPR, etc... when I read this PR title. [Duckduckgo](https://duckduckgo.com/?q=data+protection) and [Google](https://www.google.com/search?q=data+protection) have top results related to similar topics. If I live in a different bubble and this is the most unambiguous term, feel free to keep it. Just wondering if there's a less overloaded term.

Comment (xing-yang) [reply wg-data-protection/charter.md]: I'm reading this article which gives a definition of data protection: https://searchdatabackup.techtarget.com/definition/data-protection.  It says "a large part of a data protection strategy is ensuring that data can be restored quickly after any corruption or loss. Protecting data from compromise and ensuring data privacy are other key components of data protection."  So the first part is what we are focusing on in this WG.  GDPR is important but just not a focus for this WG at this point (I assume backup vendors will have to comply with GDPR in Europe, but they probably don't have to discuss about a solution in this WG).

**Comment by tallclair** [comment]
It seems like the main focus of this working group is around backup and restore (as @mikedanese mentioned, I was expecting more of a policy & compliance angle). If that is indeed the focus, can you explain why this needs to be a working group, and isn't just a subproject of SIG-Storage?

If the policy aspects are within scope, then you should also sync up with the [policy working group](https://github.com/kubernetes/community/tree/master/wg-policy) and [CNCF SIG-Security](https://github.com/cncf/sig-security), which are looking at similar issues.

**Comment by xing-yang** [comment]
> It seems like the main focus of this working group is around backup and restore (as @mikedanese mentioned, I was expecting more of a policy & compliance angle). If that is indeed the focus, can you explain why this needs to be a working group, and isn't just a subproject of SIG-Storage?

If the policy aspects are within scope, then you should also sync up with the policy working group and CNCF SIG-Security, which are looking at similar issues.

Hi @tallclair, this WG is looking at things from SIG-Apps angle as well, not just within SIG-Storage scope.

File: wg-data-protection/charter.md (hunk 18 above)
Comment (mikedanese) [reply wg-data-protection/charter.md]: I'm not asking if GDPR will be specifically addressed. I'm pointing out GDPR as a very well known example usage of "data protection"  that is primarily concern is security and privacy. If you do not plan to address these aspects of data protection, then "data loss prevention" or "storage continuity and disaster" recovery would be more accurate and less ambiguous.

Comment (xing-yang) [reply wg-data-protection/charter.md]: This WG is covering more than "storage" though as described in the "Scope" section.  Let me think about it and get back to you.

Comment (prydonius) [reply wg-data-protection/charter.md]: Agree that the term can be confused with information security, but as Xing says, it is also a valid term for Backup and Disaster Recovery and is currently used by the vendors that this group intends to attract. It's difficult to come up with another all-encompassing, well known name. We could make it more clear by renaming this to "Data Protection and Disaster Recovery", but it feels a little too long for a WG name. We can also clarify the distinction better in the charter.

Comment (xing-yang) [reply wg-data-protection/charter.md]: Added the following to clarify this: 
https://github.com/kubernetes/community/pull/4301/commits/64f4546ce58074f1c1f2a05f9e5a86cde99615df

**Comment by nrb** [comment]
>If that is indeed the focus, can you explain why this needs to be a working group, and isn't just a subproject of SIG-Storage?

From my conversations around this, I believe the reasoning includes (but isn't limited to) the following topics, which I think this document should mention explicitly:

* How are application representations inside the Kubernetes API server/etcd backed up and restored? I work on Velero, which has one approach to this, but there's also the [Stateful Application Data Management API KEP](https://github.com/kubernetes/enhancements/pull/1051)
* How is application quiescing handled before and after snapshots? This is something sig-storage has deferred to sig-apps on as I understand, with work such as [ExecutionHooks](https://github.com/kubernetes-sigs/execution-hook/pull/3)

**Comment by mikedanese** [comment]
> I am okay with SIG Auth not being explicitly listed as a stakeholder here (though I would like opinions from the other tech leads). We try to avoid being SIG "Security" since everything has some security aspect.

I agree with this assessment.

File: wg-data-protection/charter.md [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+* Design discussions on specific topics related to data protection and disaster recovery support.
+* Document results of discussions and investigations in a linkable medium.
+
+Potential design topics include, but are not limited to the following:
+* Read data from a snapshot without creating a new volume.
+* Volume backups
+* Data populator
+* Retrieve diffs between two snapshots (block and file level)
+* Consistency volume groups (group snapshot)
+* Application snapshot, backup, and recovery
```
Comment (vishnuitta) [inline wg-data-protection/charter.md]: As recovery can usually take time based on amount of data/speed etc, worth to add application migration as well in the scope.. consider the usecase like application upgrades

File: wg-data-protection/charter.md [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+* Data protection policy (Data protection policy usually means we can set up a schedule to do
```
Comment (vishnuitta) [inline wg-data-protection/charter.md]: Does this also include QoS related metrics like backup IOPS so that application side IOs won't get impacted

File: wg-data-protection/charter.md (hunk 18 above)
Comment (timothysc) [reply wg-data-protection/charter.md]: Can you add some prose with links spelling out the definition, I think that would help alot.

File: wg-data-protection/charter.md (hunk 19 above)
Comment (xing-yang) [reply wg-data-protection/charter.md]: Good point. I agree that application migration may be involved during recovery.  I'm hesitant to add it explicitly in the charter though because "migration" itself has a very wide scope and it can be used in other non-recovery related cases as well.  Application upgrade is another big topic of its own.  

What we have listed here are high level potential topics.  As a WG, we can discuss in detail how to design and implement these features.

File: wg-data-protection/charter.md (hunk 20 above)
Comment (xing-yang) [reply wg-data-protection/charter.md]: It might.  That's also something we should discuss and decide in the WG.

File: wg-data-protection/charter.md (hunk 18 above)
Comment (xing-yang) [reply wg-data-protection/charter.md]: Added some text to explain the purpose of Data Protection.  Hope that helps.  Thanks.  https://github.com/kubernetes/community/pull/4301/commits/a3d2690f20b3ef745d90939c04816bb212d2e051

File: wg-data-protection/charter.md (hunk 16 above)
Comment (xing-yang) [reply wg-data-protection/charter.md]: Statement is modified to address Derek's comments.  Thanks.

**Comment by xing-yang** [comment]
Rebased.

File: wg-data-protection/charter.md (hunk 20 above)
Comment (vishnuitta) [reply wg-data-protection/charter.md]: thanks @xing-yang for the responses.

**Comment by derekwaynecarr** [comment]
i had been out of office, but for posterity, this is lgtm

/lgtm

**Comment by xing-yang** [comment]
Thanks everyone!
